Firmware/power_panic.cpp

```diff
@@ -0,0 +1,490 @@
+    wdt_enable(WDTO_1S);
+    WRITE(BEEPER,HIGH);
+    lcd_clear();
+    lcd_puts_at_P(0, 1, MSG_POWERPANIC_DETECTED);
+    while(1);
+}
+
+
+void uvlo_() {
+    unsigned long time_start = _millis();
+
+    // True if a print is already saved to RAM
+    bool sd_print_saved_in_ram = saved_printing && (saved_printing_type == PRINTING_TYPE_SD);
+
+    // Flag to decide whether or not to set EEPROM_UVLO bit
+    bool sd_print = card.sdprinting || sd_print_saved_in_ram;
+    const bool pos_invalid = mesh_bed_leveling_flag || homing_flag;
+
+    // Conserve as much power as soon as possible
+    // Turn off the LCD backlight
+#ifdef LCD_BL_PIN
+    backlightMode = BACKLIGHT_MODE_DIM;
+    backlightLevel_LOW = 0;
+    backlight_update();
+#endif //LCD_BL_PIN
+
+    // Disable X and Y motors to conserve power
+    disable_x();
+    disable_y();
+
+    // Minimise Z and E motor currents (Hold and Run)
+#ifdef TMC2130
+    tmc2130_set_current_h(Z_AXIS, 20);
+    tmc2130_set_current_r(Z_AXIS, 20);
+    tmc2130_set_current_h(E_AXIS, 20);
+    tmc2130_set_current_r(E_AXIS, 20);
+#endif //TMC2130
+
+    if (!sd_print_saved_in_ram && !isPartialBackupAvailable)
+    {
+        saved_bed_temperature = target_temperature_bed;
+        saved_extruder_temperature = target_temperature[active_extruder];
+        saved_extruder_relative_mode = axis_relative_modes & E_AXIS_MASK;
+        saved_fan_speed = fanSpeed;
+        memcpy(saved_pos, current_position, sizeof(saved_pos));
+        if (pos_invalid) saved_pos[X_AXIS] = X_COORD_INVALID;
+    }
+
+    // Stop all heaters before continuing
+    setTargetHotend(0);
+    setTargetBed(0);
+
+    if (!sd_print_saved_in_ram) {
+        // Calculate the file position, from which to resume this print.
+        save_print_file_state();
+
+        // save the global state at planning time
+        save_planner_global_state();
+    }
+
+    // From this point on and up to the print recovery, Z should not move during X/Y travels and
+    // should be controlled precisely. Reset the MBL status before planner_abort_hard in order to
+    // get the physical Z for further manipulation.
+    bool mbl_was_active = mbl.active;
+    mbl.active = false;
+
+    // After this call, the planner queue is emptied and the current_position is set to a current logical coordinate.
+    // The logical coordinate will likely differ from the machine coordinate if the skew calibration and mesh bed leveling
+    // are in action.
+    planner_abort_hard();
+
+    // Store the print logical Z position, which we need to recover (a slight error here would be
+    // recovered on the next Gcode instruction, while a physical location error would not)
+    float logical_z = saved_pos[Z_AXIS];
+    if(mbl_was_active) {
+        // Mesh bed leveling was being actively applied to the Z-position. Revert the
+        // mesh bed leveling offset value.
+        logical_z -= mbl.get_z(st_get_position_mm(X_AXIS), st_get_position_mm(Y_AXIS));
```

Comment:
I suspect this should be `mbl.get_z(saved_pos[X_AXIS], saved_pos[Y_AXIS])` since if the extruder had been parked prior to the power panic, the X and Y machine positions are not correct (i.e. they are not equal to the saved print position)